Mathlib/RingTheory/DividedPowers/Basic.lean

```diff
@@ -0,0 +1,514 @@
+
+* `DividedPowers.dpow_null` asserts that `dpow n x = 0` for `x ∉ I`
+
+* `DividedPowers.dpow_mem` : `dpow n x ∈ I` for `n ≠ 0`
+For `x y : A` and `m n  : ℕ` such that `x ∈ I` and `y ∈ I`, one has
+* `DividedPowers.dpow_zero` : `dpow 0 x = 1`
+* `DividedPowers.dpow_one` : `dpow 1 x = 1`
+* `DividedPowers.dpow_add` : `dpow n (x + y) =
+(antidiagonal n).sum fun k => dpow k.1 x * dpow k.2 y`,
+this is the binomial theorem without binomial coefficients.
+* `DividedPowers.dpow_smul`: `dpow n (a * x) = a ^ n * dpow n x`
+* `DividedPowers.dpow_mul` : `dpow m x * dpow n x = choose (m + n) m * dpow (m + n) x`
+* `DividedPowers.dpow_comp` : `dpow m (dpow n x) = mchoose m n * dpow (m * n) x`
+
+* `DividedPowers.dividedPowersBot` : the trivial divided powers structure on the zero ideal
+
+* `DividedPowers.prod_dpow`: a product of divided powers is a multinomial coefficients
+times a divided power
+
+* `DividedPowers.dpow_sum`: the multinomial theorem for divided powers,
+without multinomial coefficients.
+
+* `DividedPowers.ofRingEquiv`: transfer divided powers along `RingEquiv`
+
+* `DividedPowers.equiv`: the equivalence `DividedPowers I ≃ DividedPowers J`,
+for `e : R ≃+* S`, and `I : Ideal R`,  `J : Ideal S` such that `I.map e = J`
+
+* `DividedPowers.exp`: the power series `Σ (dpow n a) X ^n`
+
+* `DividedPowers, exp_add`: its multiplicativity
+
+## References
+
+* [P. Berthelot (1974), *Cohomologie cristalline des schémas de
+caractéristique $p$ > 0*][Berthelot-1974]
+
+* [P. Berthelot and A. Ogus (1978), *Notes on crystalline
+cohomology*][BerthelotOgus-1978]
+
+* [N. Roby (1963), *Lois polynomes et lois formelles en théorie des
+modules*][Roby-1963]
+
+* [N. Roby (1965), *Les algèbres à puissances dividées*][Roby-1965]
+
+## Discussion
+
+* In practice, one often has a single such structure to handle on a given ideal,
+but several ideals on the same ring might be considered.
+Without any explicit mention of the ideal, it is not clear whether such structures
+should be provided as instances.
+
+* We do not provide any notation such as `a ^[n]` for `dpow a n`.
+
+-/
+
+
+section DividedPowersDefinition
+/- ## Definition of divided powers -/
+
+open Finset Nat
+
+/-- The divided power structure on an ideal I of a commutative ring A -/
+@[ext]
+structure DividedPowers {A : Type*} [CommSemiring A] (I : Ideal A) where
+  /-- The divided power function underlying a divided power structure -/
+  dpow : ℕ → A → A
+  dpow_null : ∀ {n x} (_ : x ∉ I), dpow n x = 0
+  dpow_zero : ∀ {x} (_ : x ∈ I), dpow 0 x = 1
+  dpow_one : ∀ {x} (_ : x ∈ I), dpow 1 x = x
+  dpow_mem : ∀ {n x} (_ : n ≠ 0) (_ : x ∈ I), dpow n x ∈ I
+  dpow_add : ∀ (n) {x y} (_ : x ∈ I) (_ : y ∈ I),
+    dpow n (x + y) = (antidiagonal n).sum fun k => dpow k.1 x * dpow k.2 y
+  dpow_smul : ∀ (n) {a : A} {x} (_ : x ∈ I),
+    dpow n (a * x) = a ^ n * dpow n x
+  dpow_mul : ∀ (m n) {x} (_ : x ∈ I),
+    dpow m x * dpow n x = choose (m + n) m * dpow (m + n) x
```

Comment:
It seems `dpow_smul` would be better named `dpow_mul`, and `dpow_mul` could be named `mul_dpow` (or `dpow_mul_dpow`).

I would think that the phrase `smul` (and the notation `•`) should completely disappear from the API.

Reply:
from the API of divided powers, or from mathlib?

Reply:
in any case, I made the change

Reply:
I am not as radical as suggesting removing `smul` from mathlib...

Reply:
Do you insist that I delete it now? the point is that I don't have enough visibility of the adaptations it would need later on.